---
title: "Visualization of the YihW phylogenetic tree and multiple sequence alignment of YihW protein homologs"
author: "Anna Rybina"
output:
  html_document: #default
    toc: true
    #theme: united
    toc_depth: 5
    toc_float: true
    #number_section: true
#editor_options:
  chunk_output_type: console

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 0. Prerequisites: installation and import of libraries

The current work was done under R version 4.2.2 (2022-10-31) and required 
the following packages:
```{r message=FALSE, warning = FALSE}
require(msa)
require(aplot)
require(dplyr)
require(ggplot2)
require(gggenes)
require(ggtree)
require(glue)
require(ggmsa)
require(gggenes)
require(treeio)
require(ggtext)
require(magrittr)
require(stats)
require(stringr)
```

## 1. Loading data 

Load newick tree and root the tree manually:
```{r message=FALSE, warning = FALSE}
# load newick tree
nwk <- 'YihW_protein.fasttree'
tree <- read.tree(nwk)

# get node id by which the tree will be rooted 
node_for_rooting <- MRCA(tree, "WP_193496802.1_GCF_014960965.1", "WP_015035635.1_GCF_001886595.1")

# reroot manually 
tree2 <- root(tree, node = node_for_rooting, edgelabel = TRUE)
```

Check the prelimenary tree:
```{r}
ggtree(tree2) +  geom_tiplab(linesize=.5, size=3) +  
  coord_cartesian(clip = 'off') + 
  theme_tree2(plot.margin=margin(6, 160, 6, 6),
              axis.text.x=element_blank(), 
              axis.ticks.x=element_blank(),
              axis.line.x=element_blank()) 
```

Load multiple sequence alignment of YihW protein homologs:
```{r}
protein_msa_file <- 'YihW_protein.mafft.fasta'
tidy_prot <- tidy_msa(protein_msa_file, start = 1, end = 70)
```


Load and preprocess tree annotation data:
```{r message=FALSE, warning = FALSE}
tree_data_file <- 'tree_data_summary.tsv'
tree_data <- read.csv(tree_data_file, sep='\t')

# select specific columns 
d <- tree_data %>% select(c('tree_node_id', 'full_organism_name', 
                            'bac_class', 'class_color')) %>% 
  distinct() %>%
  filter(., tree_node_id %in% tree2$tip.label) %>% 
  dplyr::rename(label="tree_node_id", label2="full_organism_name")

# add species name 
d$species <- d$label2 %>% 
  lapply(function(x){ifelse(is.na(stringr::word(x, 1, 2)), 
                            x, 
                            stringr::word(x, 1, 2))}) %>%
  unlist() %>%  
  lapply(., function(y){stringr::str_replace(y, 'sp.', '')})

# add column specifying part of organism name without its species name
d$rest <- paste0(
  ifelse(grepl("sp.", d$label2),'sp. ', ''),
  d$label2 %>% lapply(function(x){ifelse(is.na(stringr::word(x, 3, -1)), 
                                         x, 
                                         stringr::word(x, 3, -1))}) %>% 
    unlist())

# indicate what part of organism name will be written in italic
d <-  d %>% mutate(lab=glue("italic('{species}')~'{rest}'"))

# manually mark the reference E. coli K-12 MG1655 strain bold 
d[d == "italic('Escherichia coli')~'str. K-12 substr. MG1655'"] <- "bolditalic('Escherichia coli')~bold('str. K-12 substr. MG1655')"

```


Load and prepare data for plotting the _yihW_ gene neighborhood on the YihW tree
```{r message=FALSE, warning = FALSE}
loci_file <- 'loci_data_summary.tsv'
loci_data <- read.csv(loci_file, sep='\t') %>% 
  select(c(tree_node_id, gene, start, end, strand)) 

# add orientation columnd based on the gene strand column
loci_data <-  loci_data %>%  
  mutate(orientation=ifelse(strand == 'forward',1, -1))     

# select node ids with reverse orientation of the yihW gene
flip_tips <- loci_data %>% 
  filter((gene=='yihW') & (orientation==-1)) %>% 
  pull(tree_node_id) 

# specify custom function that can mutate and replace several columns on a subset of rows
mutate_cond <- function(.data, condition, ..., envir = parent.frame()) {
  condition <- eval(substitute(condition), .data, envir)
  .data[condition, ] <- .data[condition, ] %>% mutate(...)
  .data
}

# modify gene coordinates for some organisms to make uniform orientation of the yihW gene 
loci_data <-  loci_data %>% mutate_cond(tree_node_id %in% flip_tips,
                                        start=-start, end=-end)

# add gene column in italic 
loci_data <- loci_data %>% mutate(gene_italic = glue('*{gene}*'))

```



Specify custom palettes:
```{r}
# Custom palette to distinguish methionine from other residues at certain position in the multiple sequence alignment
nonMet_color <-   '#C1AEFC'
my_cutstom <- data.frame(names = c(LETTERS[1:12], LETTERS[13], LETTERS[14:26]), 
                         color = c(rep(nonMet_color, length(LETTERS[1:12])),
                                   '#FFFFFF',
                                   rep(nonMet_color, length(LETTERS[14:26]))), 
                         stringsAsFactors = FALSE)

# Custom palette to color genes in the yihW gene neighborhood and terminal leaves in the tree 
class_gene_colors <- c(setNames(c(unique(d$class_color), '#FFFFFF'), unique(d$bac_class)), 
                        c( 'ompL'= '#c6ff00', 'yihO'= '#26DC96', 'yihP'= '#1fada2',
                           'yihQ'= '#FDF434', 'yihR'= '#ffaafe', 'yihS'= '#651fff',
                           'yihT'= '#ffc400', 'yihU'= '#d500f9', 'yihV'= '#00e5ff',
                           'yihW'= '#3d5afe', 'other'='#E0E0E0'))

# Custom palette to label color ranges 
class_gene_colors_it <- c(setNames(c(unique(d$class_color), '#FFFFFF'), unique(d$bac_class)),
                           c('*ompL*'= '#c6ff00', '*yihO*'= '#26DC96', '*yihP*'= '#1fada2',
                             '*yihQ*'= '#FDF434', '*yihR*'= '#ffaafe', '*yihS*'= '#651fff',
                             '*yihT*'= '#ffc400', '*yihU*'= '#d500f9', '*yihV*'= '#00e5ff',
                             '*yihW*'= '#3d5afe', 
                             '*other*'='#E0E0E0'))
 
```



## 2. Plotting the YihW phylogenetic tree 

Plot phylogenetic tree of YihW protein homologs
```{r message=FALSE, warning = FALSE}
# save plot as png file at the current directory
png(filename = paste0(getwd(), '/YihW_tree.png'), width=5750, height=5000,res=500)
par(mar=c(3,3,2,2)+0.1)

p <- ggtree(tree2)  +  
  coord_cartesian(clip = 'off') + 
  theme_tree2(plot.margin=margin(6, 360, 6, 6),
              axis.text.x=element_blank(), 
              axis.ticks.x=element_blank(),
              axis.line.x=element_blank()) +
  geom_treescale(x=0, y=25, fontsize=3, linesize=0.3, offset=1)

p1 <- p %<+% d + geom_tiplab(aes(label=lab, 
                                 fill = bac_class),  
                             parse=T, 
                             linesize=.5, 
                             size=4, 
                             label.size = 0, 
                             align = T, 
                             geom='label',
                             label.padding = unit(0, "lines")) + 
  geom_nodelab(aes(label=label, 
                   subset = !is.na(as.numeric(label)) & as.numeric(label) > 0.8),
               color="black", 
               size=2, 
               hjust=-.2) +
  labs(fill = "Color ranges") +
  theme(legend.position = c(0.15, 0.7),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_markdown()) +
  scale_fill_manual(
    values =unname(class_gene_colors),
    breaks=names(class_gene_colors),
    labels=names(class_gene_colors_it)
  )+
  theme(strip.text=element_blank(),
        panel.spacing=unit(10, 'cm')) 

plot(p1)
dev.off()
```



Plot the YihW protein tree together with the part of multiple sequence alignment (MSA):
```{r message=FALSE, warning = FALSE}
# save plot as png file at the current directory
png(filename = paste0(getwd(), '/YihW_tree_MSA.png'), width=11500, height=5000,res=500)
par(mar=c(3,3,2,2)+0.1)

p <- ggtree(tree2)  +  
  coord_cartesian(clip = 'off') + 
  theme_tree2(plot.margin=margin(6, 360, 6, 6),
              axis.text.x=element_blank(), 
              axis.ticks.x=element_blank(),
              axis.line.x=element_blank()) 

p2 <- p %<+% d + geom_tiplab(aes(label=lab), 
                             parse=T, 
                             linesize=.5, 
                             size=4, 
                             label.size = 0, 
                             align = T, 
                             geom='label',
                             label.padding = unit(0, "lines")) + 
  geom_nodelab(aes(label=label, 
                   subset = !is.na(as.numeric(label)) & as.numeric(label) > 0.8),
               color="black", 
               size=2, 
               hjust=-.2) +
  geom_facet(
    data=tidy_prot, 
    geom=geom_msa,
    position_highlight = c(53),
    none_bg = TRUE,
    custom_color=my_cutstom, 
    panel = 'msa',
    char_width = 0.8
  ) +
  theme(strip.text=element_blank(),
        panel.spacing=unit(10, 'cm')) 

plot(p2)
dev.off()

```



Plot the YihW protein tree together with the _yihW_ gene neighborhood: 
```{r message=FALSE, warning = FALSE}
# save plot as png file at the current directory
png(filename = paste0(getwd(), '/YihW_tree_loci.png'), width=11500, height=5000,res=500)
par(mar=c(3,3,2,2)+0.1)
p <- ggtree(tree2)  +  
  coord_cartesian(clip = 'off') + 
  theme_tree2(
    plot.margin=margin(6, 160, 6, 6),
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    axis.line.x=element_blank()) +
  geom_treescale(x=0, y=45, fontsize=3, linesize=0.3, offset=1)

p3 <- p %<+% d + geom_tiplab(aes(label=lab, 
                                 fill=bac_class),  
                             parse=T, 
                             linesize=.5, 
                             size=4, 
                             label.size = 0,
                             align = T, 
                             geom='label',
                             label.padding = unit(0, "lines")
                             ) + 
  geom_nodelab(aes(label=label, 
                   subset = !is.na(as.numeric(label)) & as.numeric(label) > 0.8), 
               color="black", 
               size=2, 
               hjust=-.2) +
  labs(fill = "Color ranges") +
  theme(legend.position = c(0.1, 0.7),
        legend.key.size = unit(0.7, 'cm'),
        legend.text = element_markdown()) +
  geom_facet(mapping = aes(xmin = start, 
                           xmax = end, 
                           fill = gene,
                           forward = orientation),
  data = loci_data, 
  geom = geom_motif, 
  panel = 'yih genes',
  on = "yihW", 
  align = 'left', 
  arrowhead_height = unit(4, "mm"), 
  arrowhead_width = unit(3, "mm"), 
  arrow_body_height= unit(3, "mm")) +
  scale_fill_manual(
    values =unname(class_gene_colors),
    breaks=names(class_gene_colors),
    labels=names(class_gene_colors_it)
  )+
  scale_x_continuous(expand=c(0,0)) +
  theme(strip.text=element_blank(),
        panel.spacing=unit(13, 'cm')) 

plot(p3)
dev.off()
```


## 3. Plotting MSA of YihW protein homologs

Load MSA containing only sequences related to Proteobacteria: 
```{r}
prot_Proteobac_only <- 'YihW_protein_Proteobac.mafft.fasta'
prot_seq <-  readAAStringSet(prot_Proteobac_only)
```

Reorder and rename protein sequences in the MSA:
```{r}
# construct plotiing "basis" for the tree
p <- ggtree(tree2)  +  
  coord_cartesian(clip = 'off') + 
  theme_tree2(
    plot.margin=margin(6, 160, 6, 6),
    axis.text.x=element_blank(), 
    axis.ticks.x=element_blank(),
    axis.line.x=element_blank()) +
  geom_treescale(x=0, y=45, fontsize=3, linesize=0.3, offset=1)

# get node ids from the tree that are present in the MSA
selected_order<- get_taxa_name(p)[get_taxa_name(p) %in% names(prot_seq)] 

# reorder sequences in the MSA according to the tree 
prot2 <- prot_seq[selected_order]

# reorder annotation data
d <- d[match(get_taxa_name(p), d$label),]

# add (1) and (2) to ids of paralogues for avoiding duplicated names in the MSA 
d$new_rest <- ave(as.character(d$rest), d$rest, FUN=function(x) if (length(x)>1) paste0(x[1], '(', seq_along(x), ')') else x[1])

# create a new annotation dataset
d2 <-  d %>% mutate(new_lab=glue("{species} {new_rest}"))

# rename protein sequences in the MSA
incorrect <- d2$label
correct <- d2$new_lab
names(correct) <- incorrect
names(prot2) = str_replace_all(names(prot2), correct, names(correct))
```


Plot part of the MSA to indicate specific residues predicted to establish polar contact with candidate effectors sulfquinovose and lactose
```{r message=FALSE, warning = FALSE}
# save plot at the current directory
png(filename = paste0(getwd(), '/YihW_MSA_pos50_110.png'), width=13500, height=6000,res=1000)
par(mar=c(0.01,3,0.01,2)+0.1)
p4 <- ggmsa(prot2, seq_name=T, start=c(50), end=c(110), char_width = 0.8,
      position_highlight = c(
        53, # Conserved Met25, candidate start of the shortened YihW product
        91, # Arg39 of the shortened YihW variant
        107 # Ser54 of the shortened YihW variant
      ),  
      color="Chemistry_AA") 
plot(p4)
dev.off()
plot(p4)
```


```{r message=FALSE, warning = FALSE}
# save plot at the current directory
png(filename = paste0(getwd(), '/YihW_MSA_pos200_255.png'), width=13500, height=6000,res=1000)
par(mar=c(0.01,3,0.01,2)+0.1)
ggmsa(prot2, seq_name=T, start=c(200), end=c(255), char_width = 0.8, 
      position_highlight = c(
        223, # Asn150 of the full-length YihW variant
        246 # Lys151 of the shortened YihW variant
      ),  
      color="Chemistry_AA")  
dev.off()
```


